tools/js_optimizer.py

```diff
@@ -239,40 +236,35 @@ def check_symbol_mapping(p):
   with ToolchainProfiler.profile_block('js_optimizer.split_to_chunks'):
     # if we are making source maps, we want our debug numbering to start from the
     # top of the file, so avoid breaking the JS into chunks
-    cores = shared.get_num_cores()
 
-    if not just_split:
-      intended_num_chunks = int(round(cores * NUM_CHUNKS_PER_CORE))
-      chunk_size = min(MAX_CHUNK_SIZE, max(MIN_CHUNK_SIZE, total_size / intended_num_chunks))
-      chunks = chunkify(funcs, chunk_size)
-    else:
+    if just_split:
       # keep same chunks as before
       chunks = [f[1] for f in funcs]
+    else:
+      intended_num_chunks = round(shared.get_num_cores() * NUM_CHUNKS_PER_CORE)
+      chunk_size = min(MAX_CHUNK_SIZE, max(MIN_CHUNK_SIZE, total_size / intended_num_chunks))
+      chunks = chunkify(funcs, chunk_size)
 
-    chunks = [chunk for chunk in chunks if len(chunk)]
-    if DEBUG and len(chunks):
+    chunks = [chunk for chunk in chunks if chunk]
+    if DEBUG:
       print('chunkification: num funcs:', len(funcs), 'actual num chunks:', len(chunks), 'chunk size range:', max(map(len, chunks)), '-', min(map(len, chunks)), file=sys.stderr)
     funcs = None
 
-    if len(chunks):
-      serialized_extra_info = ''
-      if minify_globals:
-        serialized_extra_info += '// EXTRA_INFO:' + json.dumps(minify_info)
-      elif extra_info:
-        serialized_extra_info += '// EXTRA_INFO:' + json.dumps(extra_info)
-      with ToolchainProfiler.profile_block('js_optimizer.write_chunks'):
-        def write_chunk(chunk, i):
-          temp_file = temp_files.get('.jsfunc_%d.js' % i).name
-          with open(temp_file, 'w') as f:
-            f.write(chunk)
-            f.write(serialized_extra_info)
-          return temp_file
-        filenames = [write_chunk(chunks[i], i) for i in range(len(chunks))]
-    else:
-      filenames = []
-
-  with ToolchainProfiler.profile_block('run_optimizer'):
-    if len(filenames):
+    serialized_extra_info = ''
+    if minify_globals:
+      assert not extra_info
+      serialized_extra_info += '// EXTRA_INFO:' + json.dumps(minify_info)
+    elif extra_info:
+      serialized_extra_info += '// EXTRA_INFO:' + json.dumps(extra_info)
+    with ToolchainProfiler.profile_block('js_optimizer.write_chunks'):
+      def write_chunk(chunk, i):
+        temp_file = temp_files.get('.jsfunc_%d.js' % i).name
+        utils.write_file(temp_file, chunk + serialized_extra_info)
```

Comment:
I'm not sure it matters.  

Internally python is probably just going to create some of rope pointing the two original strings and then write that.

Also, if these chunks are less than a few 10s of mb each I don't think the performance of this operation is going to have any bearing on the performance of the js optimizer phase.  Imagine the code of parsing and then serializing 10mb of JS compared to the cost of python making a copy of a 10mb string.

Reply:
> Internally python is probably just going to create some of rope pointing the two original strings and then write that.

I believe that's not the case, actually. First, this was a big performance issue in the asm.js days - we had huge strings in python (all the asm.js of Unreal Engine!) and it was quite slow until we started to write them more efficiently like this. Second, I did some searching now and can't find anyone mentioning if Python uses ropes internally, but there are lots of rope libraries for Python, which suggests to me that it still doesn't.

Now, that was huge asm.js back in the day, and maybe now the file sizes are much smaller. Except perhaps for wasm2js? So that might be worth measuring.

Reply:
OK I will measure.  We have also upgrade our python since the old days too.

Reply:
I added a stress test for this chunk writing code which creates 50Mb JS file and then runs minifyWhitespace on it and it doesn't look like there is any measurable difference here:

Before:                                                                                               
                
``` 
run1:                                           
block "js_optimizer.write_chunks" took 0.089 seconds                                
block "run_optimizer" took 2.719 seconds          
                                
run2:                                                                                 
block "js_optimizer.write_chunks" took 0.090 seconds                                
block "run_optimizer" took 2.747 seconds                                            
```
                                                                                    
After:                                                                              
 
```
run1:
block "js_optimizer.write_chunks" took 0.090 seconds                                
block "run_optimizer" took 2.740 seconds                                            
                                                                                    
run2:                                                                                    
block "js_optimizer.write_chunks" took 0.094 seconds                                
block "run_optimizer" took 2.723 seconds  
```

Here I'm measuring the cost of writing the chunks against the cost of running the optimizer both before and after this change.

Reply:
BTW. writing that test and running it showed that I had actually broken the chunkification code with my initial version of this PR... I guess we don't have good tests that verify that chunkification is really happening?

Reply:
Thanks for measuring, sounds good.

We have `test_js_optimizer_chunk_size_determinism` but I guess that checks determinism, and not that we actually chunk. Maybe it would be good to add a test here?

Reply:
Its hard because the chunking is really an implemenation detail so it tricky to test with our normal black-box testing approach.   Maybe some unit tests for js_optimizer.py itself?

Reply:
I don't think we need to block this change on that though..

Reply:
Yeah, I guess we don't need to block on that, since you measured compile time are ok.